contracts/contracts/rateX/RateX.sol

```diff
@@ -1,222 +1,235 @@
+  ///@param _tokenIn Address of token we are swapping from
+  ///@param _tokenOut Address of token we are swapping to
+  ///@param _amountIn Amount of tokenIn for swap
+  ///@param _quotedAmountWithSlippageProtection min amount of tokenOut we want to get
+  ///@param _recipient Address of recipient
+  ///@return amountOut Amount of tokenOut we got
+  ///@dev Right now we don't have any optimizations for gas costs and calldata size
+  function swap(
+    Route[] calldata _foundRoutes,
+    address _tokenIn,
+    address _tokenOut,
+    uint256 _amountIn,
+    uint256 _quotedAmountWithSlippageProtection,
+    address _recipient
+  ) external nonReentrant returns (uint256 amountOut) {
+    if (_foundRoutes.length == 0) {
+      revert RateX__NoRoutes();
     }
 
-    function swapOnOneRoute(Route calldata _route)
-        internal returns(uint256 amountOut)
-    {
-        amountOut = _route.amountIn;
+    _checkAmountIn(_foundRoutes, _amountIn);
+
+    TransferHelper.safeTransferFrom(_tokenIn, msg.sender, address(this), _amountIn);
 
-        for (uint256 i = 0; i < _route.swaps.length; ++i) {
-            SwapStep memory swapStep = _route.swaps[i];
+    uint256 balanceBefore = IERC20(_tokenOut).balanceOf(address(this));
+    amountOut = _swapForTotalAmountOut(_foundRoutes);
+    uint256 balanceAfter = IERC20(_tokenOut).balanceOf(address(this));
 
-            require(dexes[swapStep.dexId] != address(0), "Dex does not exist");
+    if (balanceAfter - balanceBefore != amountOut) {
+      revert RateX__AmountOutDoesNotMatch();
+    }
+
+    if (amountOut < _quotedAmountWithSlippageProtection) {
+      revert RateX__AmountLesserThanMinAmount();
+    }
 
-            TransferHelper.safeApprove(swapStep.tokenIn, dexes[swapStep.dexId], amountOut);
+    TransferHelper.safeTransfer(_tokenOut, _recipient, amountOut);
 
-            amountOut = IDex(dexes[swapStep.dexId]).swap(
-                swapStep.poolId,
-                swapStep.tokenIn,
-                swapStep.tokenOut,
-                amountOut,
-                0,
-                address(this)
-            );
-        }
+    emit SwapEvent(_tokenIn, _tokenOut, _amountIn, amountOut, _recipient);
+  }
+
+  function _swapForTotalAmountOut(Route[] calldata _foundRoutes) internal returns (uint256 amountOut) {
+    amountOut = 0;
+    for (uint256 i = 0; i < _foundRoutes.length; ++i) {
+      amountOut += _swapOnOneRoute(_foundRoutes[i]);
+    }
+  }
+
+  function _swapOnOneRoute(Route calldata _route) internal returns (uint256 amountOut) {
+    amountOut = _route.amountIn;
+
+    for (uint256 i = 0; i < _route.swaps.length; ++i) {
+      SwapStep memory swapStep = _route.swaps[i];
+
+      if (dexes[swapStep.dexId] == address(0)) {
+        revert RateX__DexDoesNotExist();
+      }
+
+      // Delegate call to the DEX contract's swap function
+      (bool success, bytes memory result) = dexes[swapStep.dexId].delegatecall(
+        abi.encodeWithSignature(
```

Comment:
More flexible approach is to send bytes for example, and every specific dex parses it to structure it needs. This way we avoid situation in which some of the dexes does't use all the params. This is also the case for sending routes off-chain, think about if it is better to avoid moving around this large arrays of structs between fn calls